File: backend/app/payroll/payslip_service.py
"""Payslip / payment-advice generation, versioning, and PDF rendering.

Generation snapshots everything the document shows (including historical
obligation balances) into an immutable JSON blob, renders a PDF via WeasyPrint,
stores it in MinIO, and supersedes any prior version for the same run+engagement.
"""
from __future__ import annotations

from datetime import datetime, timezone

from sqlalchemy.orm import Session

from app.models.enums import EngagementType
from app.models.organization import Organization
from app.models.payroll import PayrollRun, PayrollRunPerson
from app.models.payslip import DocumentType, Payslip
from app.models.person import Engagement, Person
from app.models.project import Project
from app.reports import pdf as pdf_renderer
from app.services import obligation_service
from app.storage import documents


def _mask_account(acct: str | None) -> str:
    if not acct:
        return "—"
    tail = acct[-4:]
    return f"{'*' * max(len(acct) - 4, 0)}{tail}"


def build_snapshot(db: Session, run: PayrollRun, line: PayrollRunPerson) -> dict:
    engagement: Engagement = db.get(Engagement, line.engagement_id)
    person: Person = db.get(Person, engagement.person_id)
    org: Organization = db.get(Organization, run.organization_id)
    period = run.period

    dept = engagement.department.name if engagement.department else None
    position = engagement.position.title if engagement.position else None
    project = None
    if engagement.project_id:
        p = db.get(Project, engagement.project_id)
        project = p.project_name if p else None

    is_consultant = engagement.engagement_type in EngagementType.CONSULTANT_TYPES
    doc_type = DocumentType.PAYMENT_ADVICE if is_consultant else DocumentType.PAYSLIP

    obligations = obligation_service.obligation_summary_as_of(
        db,
        engagement_id=engagement.id,
        person_id=person.id,
        payroll_run_id=run.id,
        period_end=period.period_end,
    )

    header = {
        "company_name": org.name if org else "",
        "organization_code": org.code if org else "",
        "payroll_period": period.name,
        "period_start": period.period_start.isoformat(),
        "period_end": period.period_end.isoformat(),
        "pay_date": period.pay_date.isoformat() if period.pay_date else None,
        "name": person.full_name,
        "employee_number": engagement.employee_number,
        "department": dept,
        "position": position,
        "employment_type": engagement.engagement_type,
        "project": project,
        "cost_center": engagement.cost_center,
        "bank_account_masked": _mask_account(person.bank_account_number),
        "bank_name": person.bank_name,
    }

    return {
        "document_type": doc_type,
        "header": header,
        "earnings": line.earnings or {},
        "gross_pay": float(line.gross_pay or 0),
        "deductions": line.deductions or {},
        "total_deductions": float(line.total_deductions or 0),
        "net_pay": float(line.net_pay or 0),
        "obligations": obligations,
        "run_reference": run.reference,
        "rule_versions": run.rule_version_snapshot or {},
        "generated_at": datetime.now(timezone.utc).isoformat(),
        "prototype_notice": (
            "PROTOTYPE — statutory values are illustrative, not authoritative."
        ),
    }


def _pdf_key(payslip_uuid: str) -> str:
    return f"payslips/{payslip_uuid}.pdf"


def generate_for_run(
    db: Session, run: PayrollRun, user_id: int | None = None, render_pdf: bool = True
) -> list[Payslip]:
    """(Re)generate documents for every line in the run.

    Existing current documents for the same engagement are superseded and a new
    version is created — old versions are preserved for audit. When `render_pdf`
    is False, records are created without pre-rendering PDFs (used by the seeder;
    PDFs are then rendered on demand from the immutable snapshot).
    """
    lines = db.query(PayrollRunPerson).filter(PayrollRunPerson.run_id == run.id).all()
    created: list[Payslip] = []

    for line in lines:
        prior = (
            db.query(Payslip)
            .filter(
                Payslip.payroll_run_id == run.id,
                Payslip.engagement_id == line.engagement_id,
                Payslip.is_current.is_(True),
            )
            .first()
        )
        version = 1
        if prior:
            prior.is_current = False
            prior.status = "SUPERSEDED"
            version = prior.version + 1

        snapshot = build_snapshot(db, run, line)
        engagement = db.get(Engagement, line.engagement_id)
        snapshot["version"] = version

        payslip = Payslip(
            organization_id=run.organization_id,
            payroll_run_id=run.id,
            engagement_id=line.engagement_id,
            person_id=engagement.person_id,
            document_type=snapshot["document_type"],
            version=version,
            is_current=True,
            status="ISSUED",
            net_pay=line.net_pay or 0,
            snapshot=snapshot,
            generated_by=user_id,
        )
        db.add(payslip)
        db.flush()  # assign uuid

        # Render + store the PDF (non-fatal if object storage is unavailable).
        if render_pdf:
            try:
                pdf_bytes = pdf_renderer.render_document(snapshot)
                key = _pdf_key(payslip.uuid)
                documents.put_bytes(key, pdf_bytes, content_type="application/pdf")
                payslip.pdf_object_key = key
            except Exception as exc:  # noqa: BLE001
                print(f"[payslip] PDF store skipped for {payslip.uuid}: {exc}")

        created.append(payslip)

    db.flush()
    return created


def generate_for_engagement(
    db: Session, run: PayrollRun, engagement_id: int, user_id: int | None = None
) -> Payslip:
    """Create (or return the current) payslip for ONE engagement in a run.

    Used by employee self-service — an employee generates their own payslip from
    a finalized run. Numbers come from the official PayrollRunPerson line.
    """
    existing = (
        db.query(Payslip)
        .filter(Payslip.payroll_run_id == run.id, Payslip.engagement_id == engagement_id,
                Payslip.is_current.is_(True))
        .first()
    )
    if existing:
        return existing

    line = (
        db.query(PayrollRunPerson)
        .filter(PayrollRunPerson.run_id == run.id, PayrollRunPerson.engagement_id == engagement_id)
        .first()
    )
    if not line:
        raise ValueError("No payroll line for this engagement in the run")

    snapshot = build_snapshot(db, run, line)
    snapshot["version"] = 1
    engagement = db.get(Engagement, engagement_id)
    payslip = Payslip(
        organization_id=run.organization_id, payroll_run_id=run.id, engagement_id=engagement_id,
        person_id=engagement.person_id, document_type=snapshot["document_type"], version=1,
        is_current=True, status="ISSUED", net_pay=line.net_pay or 0, snapshot=snapshot,
        generated_by=user_id,
    )
    db.add(payslip)
    db.flush()
    return payslip


def get_pdf_bytes(db: Session, payslip: Payslip) -> bytes:
    """Return the PDF, from MinIO if stored else re-rendered from the snapshot."""
    if payslip.pdf_object_key:
        data = documents.get_bytes(payslip.pdf_object_key)
        if data:
            return data
    return pdf_renderer.render_document(payslip.snapshot)

File: bluehost/public/app/PayslipService.php
<?php
// Payslip snapshot build, versioned generation, and printable HTML rendering.

class PayslipService
{
    const CONSULTANT_TYPES = ['CONSULTANT_INDIVIDUAL', 'CONSULTANT_COMPANY'];

    public static function buildSnapshot(array $run, array $line): array
    {
        $eng = Database::one('SELECT * FROM engagements WHERE id = ?', [$line['engagement_id']]);
        $person = Database::one('SELECT * FROM people WHERE id = ?', [$eng['person_id']]);
        $org = Database::one('SELECT * FROM organizations WHERE id = ?', [$run['organization_id']]);
        $period = Database::one('SELECT * FROM payroll_periods WHERE id = ?', [$run['period_id']]);
        $dept = $eng['department_id'] ? Database::scalar('SELECT name FROM departments WHERE id = ?', [$eng['department_id']]) : null;
        $pos = $eng['position_id'] ? Database::scalar('SELECT title FROM positions WHERE id = ?', [$eng['position_id']]) : null;
        $proj = $eng['project_id'] ? Database::scalar('SELECT project_name FROM projects WHERE id = ?', [$eng['project_id']]) : null;
        $isCon = in_array($eng['engagement_type'], self::CONSULTANT_TYPES, true);

        $name = trim(implode(' ', array_filter([$person['first_name'], $person['middle_name'], $person['last_name'], $person['suffix']])));
        $obl = Obligation::summaryAsOf((int) $eng['id'], (int) $person['id'], (int) $run['id'], $period['period_end']);

        return [
            'document_type' => $isCon ? 'PAYMENT_ADVICE' : 'PAYSLIP',
            'header' => [
                'company_name' => $org['name'], 'organization_code' => $org['code'],
                'payroll_period' => $period['name'], 'period_start' => $period['period_start'],
                'period_end' => $period['period_end'], 'pay_date' => $period['pay_date'],
                'name' => $name, 'employee_number' => $eng['employee_number'],
                'department' => $dept, 'position' => $pos, 'employment_type' => $eng['engagement_type'],
                'project' => $proj, 'cost_center' => $eng['cost_center'],
                'bank_account_masked' => Util::maskAccount($person['bank_account_number']),
                'bank_name' => $person['bank_name'],
            ],
            'earnings' => json_decode($line['earnings'] ?: '{}', true),
            'gross_pay' => (float) $line['gross_pay'],
            'deductions' => json_decode($line['deductions'] ?: '{}', true),
            'total_deductions' => (float) $line['total_deductions'],
            'net_pay' => (float) $line['net_pay'],
            'obligations' => $obl,
            'run_reference' => $run['reference'],
            'rule_versions' => json_decode($run['rule_version_snapshot'] ?: '{}', true),
            'prototype_notice' => 'Statutory contributions computed using the 2025 SSS, PhilHealth, Pag-IBIG and BIR (TRAIN) withholding tables. Verify against the latest official circulars.',
        ];
    }

    /** (Re)generate one engagement's current payslip in a run; returns the payslip row. */
    public static function generateForEngagement(array $run, int $engagementId, ?int $userId): array
    {
        $existing = Database::one(
            'SELECT * FROM payslips WHERE payroll_run_id = ? AND engagement_id = ? AND is_current = 1',
            [$run['id'], $engagementId]);
        if ($existing) return $existing;

        $line = Database::one('SELECT * FROM payroll_run_people WHERE run_id = ? AND engagement_id = ?',
            [$run['id'], $engagementId]);
        if (!$line) throw new HttpError('No payroll line for this engagement', 404);

        $snap = self::buildSnapshot($run, $line);
        $snap['version'] = 1;
        $eng = Database::one('SELECT person_id FROM engagements WHERE id = ?', [$engagementId]);
        $id = Database::insert('payslips', [
            'uuid' => Util::uuid(), 'organization_id' => $run['organization_id'], 'payroll_run_id' => $run['id'],
            'engagement_id' => $engagementId, 'person_id' => $eng['person_id'], 'document_type' => $snap['document_type'],
            'version' => 1, 'is_current' => 1, 'status' => 'ISSUED', 'net_pay' => $line['net_pay'],
            'snapshot' => json_encode($snap), 'generated_by' => $userId,
        ]);
        return Database::one('SELECT * FROM payslips WHERE id = ?', [$id]);
    }

    public static function generateForRun(array $run, ?int $userId): int
    {
        $lines = Database::all('SELECT * FROM payroll_run_people WHERE run_id = ?', [$run['id']]);
        $count = 0;
        foreach ($lines as $line) {
            $prior = Database::one('SELECT * FROM payslips WHERE payroll_run_id = ? AND engagement_id = ? AND is_current = 1',
                [$run['id'], $line['engagement_id']]);
            $version = 1;
            if ($prior) {
                Database::update('payslips', (int) $prior['id'], ['is_current' => 0, 'status' => 'SUPERSEDED']);
                $version = (int) $prior['version'] + 1;
            }
            $snap = self::buildSnapshot($run, $line);
            $snap['version'] = $version;
            $eng = Database::one('SELECT person_id FROM engagements WHERE id = ?', [$line['engagement_id']]);
            Database::insert('payslips', [
                'uuid' => Util::uuid(), 'organization_id' => $run['organization_id'], 'payroll_run_id' => $run['id'],
                'engagement_id' => $line['engagement_id'], 'person_id' => $eng['person_id'],
                'document_type' => $snap['document_type'], 'version' => $version, 'is_current' => 1, 'status' => 'ISSUED',
                'net_pay' => $line['net_pay'], 'snapshot' => json_encode($snap), 'generated_by' => $userId,
            ]);
            $count++;
        }
        return $count;
    }

    private static function money($v): string { return 'PHP ' . number_format((float) $v, 2); }

    /** Render a printable HTML payslip (browser → Save as PDF on shared hosting). */
    public static function renderHtml(array $snap): string
    {
        $h = $snap['header'];
        $rows = fn($items) => implode('', array_map(
            fn($k, $v) => '<tr><td>' . ucwords(str_replace('_', ' ', $k)) . '</td><td class="n">' . self::money($v) . '</td></tr>',
            array_keys($items), array_values($items)));
        $oblRows = '';
        foreach ($snap['obligations']['rows'] as $o) {
            $oblRows .= '<tr><td>' . htmlspecialchars($o['description']) . '</td><td class="n">' . self::money($o['original_amount'])
                . '</td><td class="n">' . self::money($o['current_deduction']) . '</td><td class="n">' . self::money($o['total_paid'])
                . '</td><td class="n">' . self::money($o['remaining_balance']) . '</td><td class="n">' . $o['remaining_installments'] . '</td></tr>';
        }
        $title = $snap['document_type'] === 'PAYMENT_ADVICE' ? 'Payment Advice' : 'Payslip';
        return '<!doctype html><html><head><meta charset="utf-8"><title>' . $title . '</title><style>
          body{font-family:Arial,Helvetica,sans-serif;color:#1f2937;font-size:12px;max-width:760px;margin:20px auto;padding:0 16px}
          h1{font-size:18px;color:#1e3a8a;margin:0} .sub{color:#6b7280;font-size:11px}
          table{width:100%;border-collapse:collapse;margin-top:6px} .amt th,.amt td{border:1px solid #e5e7eb;padding:5px 7px;text-align:left}
          .amt th{background:#f3f4f6} .n{text-align:right} .kv td{padding:2px 4px} .kv td.k{color:#6b7280;width:130px}
          .net{font-size:15px;color:#065f46;font-weight:bold} .section{margin:14px 0 2px;font-weight:bold;border-bottom:2px solid #e5e7eb;padding-bottom:2px}
          .notice{background:#fff7ed;color:#9a3412;padding:5px 8px;font-size:10px;border:1px solid #fed7aa;border-radius:4px;margin-top:10px}
          @media print{.noprint{display:none}}
          </style></head><body>
          <div class="noprint" style="text-align:right;margin-bottom:8px"><button onclick="window.print()">Print / Save as PDF</button></div>
          <h1>' . htmlspecialchars($h['company_name']) . '</h1>
          <div class="sub">' . $title . ' — ' . htmlspecialchars($h['payroll_period']) . ' (' . $h['period_start'] . ' to ' . $h['period_end'] . ')</div>
          <table class="kv">
            <tr><td class="k">Employee</td><td>' . htmlspecialchars($h['name']) . '</td><td class="k">Employee No.</td><td>' . htmlspecialchars($h['employee_number'] ?? '—') . '</td></tr>
            <tr><td class="k">Department</td><td>' . htmlspecialchars($h['department'] ?? '—') . '</td><td class="k">Position</td><td>' . htmlspecialchars($h['position'] ?? '—') . '</td></tr>
            <tr><td class="k">Type</td><td>' . htmlspecialchars($h['employment_type']) . '</td><td class="k">Bank</td><td>' . htmlspecialchars(($h['bank_name'] ?? '') . ' ' . $h['bank_account_masked']) . '</td></tr>
          </table>
          <div style="display:flex;gap:14px;margin-top:8px">
            <div style="flex:1"><div class="section">Earnings</div><table class="amt"><tr><th>Item</th><th class="n">Amount</th></tr>' . $rows($snap['earnings']) . '<tr><td><b>Gross</b></td><td class="n"><b>' . self::money($snap['gross_pay']) . '</b></td></tr></table></div>
            <div style="flex:1"><div class="section">Deductions</div><table class="amt"><tr><th>Item</th><th class="n">Amount</th></tr>' . $rows($snap['deductions']) . '<tr><td><b>Total</b></td><td class="n"><b>' . self::money($snap['total_deductions']) . '</b></td></tr></table></div>
          </div>
          <table class="amt" style="margin-top:10px"><tr><td><b>NET PAY</b></td><td class="n net">' . self::money($snap['net_pay']) . '</td></tr></table>
          <div class="section">Outstanding Obligations (as of ' . $h['period_end'] . ')</div>'
          . ($oblRows ? '<table class="amt"><tr><th>Description</th><th class="n">Original</th><th class="n">This Period</th><th class="n">Total Paid</th><th class="n">Remaining</th><th class="n">Inst. Left</th></tr>' . $oblRows
             . '<tr><td colspan="4"><b>Total Outstanding</b></td><td class="n"><b>' . self::money($snap['obligations']['total_outstanding']) . '</b></td><td></td></tr></table>'
             : '<div class="sub">No outstanding obligations as of this period.</div>')
          . (!empty($snap['prototype_notice']) ? '<div class="notice">' . htmlspecialchars($snap['prototype_notice']) . '</div>' : '')
          . '</body></html>';
    }
}
